java/ql/src/semmle/code/java/security/RequestForgery.qll

```diff
@@ -0,0 +1,334 @@
+      this.asExpr() = ma.getArgument(0)
+    )
+  }
+}
+
+/**
+ * An argument to a Spring `RestTemplate` method call taken as a
+ *   sink for request forgery vulnerabilities.
+ */
+private class SpringRestTemplateArgument extends RequestForgerySink {
+  SpringRestTemplateArgument() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getMethod().(SpringRestTemplateUrlMethod).getUrlArgument(ma)
+    )
+  }
+}
+
+/**
+ * An argument to a `javax.ws.rs.Client` `target` method call taken as a
+ *   sink for request forgery vulnerabilities.
+ */
+private class JaxRsClientTarget extends RequestForgerySink {
+  JaxRsClientTarget() {
+    exists(MethodAccess ma |
+      ma.getMethod().getDeclaringType() instanceof JaxRsClient and
+      ma.getMethod().hasName("target")
+    |
+      this.asExpr() = ma.getArgument(0)
+    )
+  }
+}
+
+/**
+ * An argument to an `org.springframework.http.RequestEntity` constructor call
+ *   which is an URI taken as a sink for request forgery vulnerabilities.
+ */
+private class RequestEntityUriArg extends RequestForgerySink {
+  RequestEntityUriArg() {
+    exists(ClassInstanceExpr e, Argument a |
+      e.getConstructedType() instanceof SpringRequestEntity and
+      e.getAnArgument() = a and
+      a.getType() instanceof TypeUri and
+      this.asExpr() = a
+    )
+  }
+}
+
+/**
+ * A Spring Rest Template method
+ * which take a URL as an argument.
+ */
+private class SpringRestTemplateUrlMethod extends Method {
+  SpringRestTemplateUrlMethod() {
+    this.getDeclaringType() instanceof SpringRestTemplate and
+    this.hasName([
+        "doExecute", "postForEntity", "postForLocation", "postForObject", "put", "exchange",
+        "execute", "getForEntity", "getForObject", "patchForObject"
+      ])
+  }
+
+  /**
+   * Gets the argument which corresponds to a URL argument
+   * passed as a `java.net.URL` object or as a string or the like
+   */
+  Argument getUrlArgument(MethodAccess ma) {
+    // doExecute(URI url, HttpMethod method, RequestCallback requestCallback,
+    //  ResponseExtractor<T> responseExtractor)
+    result = ma.getArgument(0)
+  }
+}
+
+/** A sanitizer for request forgery vulnerabilities. */
+abstract class RequestForgerySanitizer extends DataFlow::Node { }
+
+private class PrimitiveSanitizer extends RequestForgerySanitizer {
+  PrimitiveSanitizer() { this.getType() instanceof PrimitiveType }
+}
+
+private class HostnameSanitizingPrefix extends CompileTimeConstantExpr {
+  int offset;
+
+  HostnameSanitizingPrefix() {
+    // Matches strings that look like when prepended to untrusted input, they will restrict
+    // the host or entity addressed: for example, anything containing `?` or `#`, or a slash that
+    // doesn't appear to be a protocol specifier (e.g. `http://` is not sanitizing), or specifically
+    // the string "/".
+    exists(
+      this.getStringValue()
+          .regexpFind(".*([?#]|[^?#:/\\\\][/\\\\]).*|[/\\\\][^/\\\\].*|^/$", 0, offset)
```

Comment:
How come backslash is included here? Isn't it just forward slash that's used in hostnames?

Reply:
I took that regex from the equivalent Go query, which was dropped in with the initial commit of `codeql-go` (https://github.com/github/codeql-go/blame/6e4880bc534a8a81819face7687a414f35534823/ql/src/semmle/go/security/UrlConcatenation.qll#L59). @max-schaefer do you recall if you had a reason for caring about backslashes here?

Reply:
I think we added that because browsers are generally pretty happy to accept backslashes in place of forward slashes. cURL, on the other hand, is pickier, so depending on the exact circumstances this may or may not be what you want.